pkg/daemon/daemon.go

```diff
@@ -119,19 +127,158 @@ func (d *daemon) Run() {
+	}
+	netAttInfo, err := d.kubeClient.GetNetworkAttachmentDefinition(networkNamespace, networkName)
+	if err != nil {
+		return "", nil, fmt.Errorf("failed to get networkName attachment %s with error: %v", networkName, err)
+	}
+
+	log.Debug().Msgf("networkName attachment %v", netAttInfo)
+	networkSpec := make(map[string]interface{})
+	err = json.Unmarshal([]byte(netAttInfo.Spec.Config), &networkSpec)
+	if err != nil {
+		return "", nil, fmt.Errorf("failed to parse networkName attachment %s with error: %v", networkName, err)
+	}
+	log.Debug().Msgf("networkName attachment spec %+v", networkSpec)
+
+	ibCniSpec, err := utils.GetIbSriovCniFromNetwork(networkSpec)
+	if err != nil {
+		return "", nil, fmt.Errorf("failed to get InfiniBand SR-IOV CNI spec from network attachment %+v, with error %v",
+			networkSpec, err)
+	}
+
+	log.Debug().Msgf("CNI spec %+v", ibCniSpec)
+	return networkName, ibCniSpec, nil
+}
+
+// Get Pod's IB network specified by name. Update netMap with Pod's networks.
+func getIbSriovPodNetwork(netName string, pw *PodWrapper, netMap map[types.UID][]*v1.NetworkSelectionElement) error {
+	var err error
+	networks, ok := netMap[pw.pod.UID]
+	if !ok {
+		networks, err = netAttUtils.ParsePodNetworkAnnotation(pw.pod)
+		if err != nil {
+			return fmt.Errorf("failed to read pod networkName annotations pod namespace %s name %s, with error: %v",
+				pw.pod.Namespace, pw.pod.Name, err)
+		}
+
+		netMap[pw.pod.UID] = networks
+	}
+	network, err := utils.GetPodNetwork(networks, netName)
+	if err != nil {
+		return fmt.Errorf("failed to get pod networkName spec %s with error: %v", netName, err)
+	}
+
+	pw.ibNetwork = network
+	pw.networks = networks
+	return nil
+}
+
+// Verify if GUID already exist for given network ID and allocates new one if not
+func (d *daemon) mapGUIDToPodNetwork(allocatedGUID, podNetworkID string, podUID types.UID) error {
+	if _, exist := d.guidPodNetworkMap[allocatedGUID]; exist {
+		if podNetworkID != d.guidPodNetworkMap[allocatedGUID] {
+			return fmt.Errorf("failed to allocate requested guid %s, already allocated for %s",
+				allocatedGUID, d.guidPodNetworkMap[allocatedGUID])
+		}
+	} else if err := d.guidPool.AllocateGUID(allocatedGUID); err != nil {
+		return fmt.Errorf("failed to allocate GUID for pod ID %s, wit error: %v", podUID, err)
+	} else {
+		d.guidPodNetworkMap[allocatedGUID] = podNetworkID
+	}
+
+	return nil
+}
+
+// Allocate GUID if network don't have it and update Pod's networks annotation
+func (d *daemon) ProceedNetworkGUID(pw *PodWrapper, networkID string) error {
+	var guidAddr guid.GUID
+	allocatedGUID, err := utils.GetPodNetworkGUID(pw.ibNetwork)
+	podNetworkID := string(pw.pod.UID) + networkID
+	if err == nil {
+		// User allocated guid manually or Pod's network was rescheduled
+		err = d.mapGUIDToPodNetwork(allocatedGUID, podNetworkID, pw.pod.UID)
+		if err != nil {
+			return err
+		}
+
+		guidAddr, err = guid.ParseGUID(allocatedGUID)
```

Comment:
I have a feeling this should be called before L#206

Reply:
I don't see a real difference where to place it. It's funny, in call of `mapGUIDToPodNetwork` there is nested call to.. `ParseGUID(allocatedGUID), so in no error cases it's called twice.

Reply:
the difference is where it would fail in case of invalid GUID.
either in `guid_pool:guidPool.AllocateGUID` or catch it early.

guid pool API kinda odd as when generating a GUID it returns a `GUID` type (uint64) but to allocate/release it you would need to convert it to a string.

Reply:
also the order here may hint that an allocation may pass but parsing the guid may fail, which is not the case